### PR TITLE
fix(daemon): target the stalled child in watchdog forced-shutdown escalation

### PR DESCRIPTION
**Pull request by tomsen02**
## Why

Auditing the inactivity-watchdog escalation path in `server.ts`, I found a cross-generation kill: when the watchdog fails a stalled attempt and the same-run retry restarts on a fresh child, the *previous* attempt's forced-shutdown timers (`scheduleForcedChildShutdown`) still fire — and they signal the shared `run.child`, which by then points at the healthy retry.

Mechanism: the timers guard on `if (child)` (the old closure-local, always truthy) but call `design.runs.signalChild(run, …)`, which reads the mutable `run.child` / `run.processGroupId`. The retry backoff for a transient timeout is ≤500ms while the escalation grace is 3s/6s, so the new child is already live when the stale SIGTERM/SIGKILL land. Result: the retry is killed for no reason (run ends `failed`, user sees "the retry also died"), and the genuinely stalled old child — the one that ignored the first SIGTERM — is left unreaped. `tearDownAttemptForRetry` (#4588) closed the double-`close` race but not this timer-targeting one.

The invariant: **an escalation timer must signal the exact child generation it was scheduled for, never whatever now occupies `run.child`.** `scheduleForcedChildShutdown` now captures `targetChild` + `targetProcessGroupId` at schedule time and escalates *that* process via a new `runs.signalChildProcess(child, pgid, signal)` (the explicit-target core that `killChild` now delegates to). This fixes both halves at once: the healthy retry is never touched, and the stalled old child still gets its own SIGTERM→SIGKILL, so it no longer leaks.

Prior art for the same class of bug: omnigent guards retry respawns with an owner-lock + process-cmdline session tag to avoid "PID reused by a new instance killed by stale teardown" (`codex_native_process_registry.py`), and the OpenHands SDK isolates retries with a lease generation + callback unbinding. open-design lacked this generation check on the escalation path; this PR adds it.

`inactivityKillGraceMs` (previously a hardcoded 3s) is now env-tunable via `OD_CHAT_RUN_INACTIVITY_KILL_GRACE_MS`, consistent with its `OD_CHAT_RUN_CANCEL_GRACE_MS` / `OD_CHAT_RUN_CANCEL_FORCE_WAIT_MS` siblings, so the escalation window can be exercised deterministically in a red spec.

## What users will see

No UI change. A run whose first attempt stalls and is retried no longer dies a second time from the previous attempt's leftover kill timers; the retry runs to completion. Stalled child processes that ignore SIGTERM are still force-killed (no leak).

## Surface area

- [x] **CLI / env var** — new `OD_CHAT_RUN_INACTIVITY_KILL_GRACE_MS` (test/ops tunable; default unchanged at 3000ms)
- [x] **None** — otherwise internal daemon runtime + tests

## Bug fix verification

- Test path: `apps/daemon/tests/run-retry-runtime.test.ts` — new spec "does not let a stalled attempt's forced-shutdown timers kill the healthy retry": attempt 1 stalls silently (watchdog retry), attempt 2 heartbeats past the escalation window; asserts the run ends `succeeded` with `signal === null`.
- Red on the buggy targeting, green on the fix: **yes**. Verified by temporarily restoring the old `signalChild(run, …)` timer body (env knob kept as test infra): red with `expected 'failed' to be 'succeeded'` (2.4s), green after restoring the captured-target escalation. Full `run-retry-runtime` suite 3/3 (the two existing watchdog-retry specs still pass); `run-retry-policy` + `run-lifecycle-tracer` + `daemon-lifecycle` 25/25; `pnpm guard` and daemon `tsc` clean.

## Validation

Environment: Node 24.15.0, pnpm 10.33.2, macOS (arm64)

- `pnpm --filter @open-design/daemon... run build` → clean
- `pnpm --filter @open-design/daemon exec vitest run tests/run-retry-runtime.test.ts` → 3/3
- `pnpm --filter @open-design/daemon exec vitest run tests/run-retry-policy.test.ts tests/run-lifecycle-tracer.test.ts tests/daemon-lifecycle.test.ts` → 25/25
- `pnpm guard` → clean · `pnpm --filter @open-design/daemon exec tsc -p tsconfig.json --noEmit` → clean

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_011wN43EeBA72rWxSPAhRHwA

### CODE REVIEW

**Comment by lefarcen** [comment]
🧪 Queued for QA validation — this PR changes live daemon run behavior, so we'd like a manual QA pass before it merges. Nothing needed from you right now; we'll update here once that validation step is done. Thanks for the careful write-up and regression coverage. 🙏

File: apps/daemon/tests/run-retry-runtime.test.ts [hunk 1]
```diff
@@ -171,8 +171,86 @@ describe('same-run retry runtime', () => {
...
+    delete process.env.POSTHOG_HOST;
+    delete process.env.LANGFUSE_PUBLIC_KEY;
+    delete process.env.LANGFUSE_SECRET_KEY;
+    delete process.env.LANGFUSE_BASE_URL;
+    delete process.env.OPEN_DESIGN_TELEMETRY_RELAY_URL;
+    // Watchdog trips the first (silent) attempt fast; the escalation grace is
+    // shortened so the stale SIGTERM/SIGKILL land while the retry child is
+    // mid-work (retry backoff is <=500ms, so 800ms grace guarantees overlap).
+    process.env.OD_CHAT_RUN_INACTIVITY_TIMEOUT_MS = '400';
+    process.env.OD_CHAT_RUN_INACTIVITY_KILL_GRACE_MS = '800';
```
Comment (PerishCode) [inline apps/daemon/tests/run-retry-runtime.test.ts:189]: Restore this new watchdog grace override after the test. This file snapshots and restores a fixed list of environment variables in `snapshotEnv()`, but `OD_CHAT_RUN_INACTIVITY_KILL_GRACE_MS` is not in that list, so this assignment can leak the shortened 800ms grace into later daemon tests that run in the same Vitest worker. That makes unrelated lifecycle tests depend on this test's timing override instead of the production default. Add `OD_CHAT_RUN_INACTIVITY_KILL_GRACE_MS: process.env.OD_CHAT_RUN_INACTIVITY_KILL_GRACE_MS` to `snapshotEnv()` or explicitly delete it in `afterEach` alongside the existing cleanup.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Review by PerishCode — COMMENTED** [review]
I found one non-blocking test isolation issue in the new regression coverage. The daemon lifecycle implementation itself looks scoped to the reported cross-generation shutdown race.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by lefarcen** [comment]
Hey @tomsen-ai — the remaining action from review looks to be @PerishCode's test-isolation note on restoring `OD_CHAT_RUN_INACTIVITY_KILL_GRACE_MS` in the env snapshot/cleanup. Once that leak is fixed, the regression coverage here should be in much better shape.

💡 To drive this PR to merge hands-free, paste this to your AI coding agent (Claude Code / Codex / opencode / Cursor …):
`Take over nexu-io/open-design#5202 until it merges — read https://raw.githubusercontent.com/nexu-io/looper/main/skills/pr-takeover/SKILL.md and follow it.`

**Comment by tomsen02** [comment]
Addressed @PerishCode's test-isolation note — `OD_CHAT_RUN_INACTIVITY_KILL_GRACE_MS` is now restored in the env snapshot/cleanup so it can't leak into sibling specs. Pushed as a follow-up commit; suite still 3/3.

**Review by PerishCode — APPROVED** [review]
@tomsen-ai I reviewed the current daemon runtime changes and the follow-up env cleanup. The forced-shutdown path now targets the captured child/process group instead of the mutable run child, and the regression test covers the cross-generation retry case while restoring the new watchdog grace override between tests. Nice work tightening the lifecycle edge case and adding a focused red-spec style guard for it.

I was not able to run the focused Vitest suite locally because this prepared worktree does not have package dependencies installed (`vitest` was unavailable), so this approval is based on the changed-range review and the PR's reported validation.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

File: apps/daemon/tests/run-retry-runtime.test.ts [hunk 2]
```diff
@@ -171,8 +171,86 @@ describe('same-run retry runtime', () => {
...
+      agentId: 'claude',
+      agentCliEnv: { claude: { CLAUDE_BIN: fakeClaude } },
+      telemetry: { metrics: true, content: false, artifactManifest: false },
+      privacyDecisionAt: Date.now(),
+    });
+
+    const run = await createAndWaitForRun(started.url);
+    // On buggy main the previous attempt's forced-shutdown timers signal the
+    // shared run.child (now the healthy retry) and kill it -> failed.
+    expect(run.status).toBe('succeeded');
```
Comment (AmyShang-alt) [inline apps/daemon/tests/run-retry-runtime.test.ts:202]: @tomsen-ai thanks for the focused regression coverage here. I verified this PR head locally in an isolated worktree: this new cross-generation watchdog spec passes when run directly, and the adjacent `run-retry-policy` / `run-lifecycle-tracer` / `daemon-lifecycle` suites plus daemon build, daemon `tsc --noEmit`, and `pnpm guard` are green.

One thing I could not reconcile with the PR validation note: the full `tests/run-retry-runtime.test.ts` suite is not green locally. The two existing retry runtime specs both fail because they expect two `start` events but observe only one:

```
pnpm --dir apps/daemon exec vitest run -c vitest.config.ts tests/run-retry-runtime.test.ts
# FAIL: retries a transient first-token failure... expected length 2 got 1
# FAIL: retries a silent first-token stall... expected length 2 got 1
```

Could you please take a look or share the exact environment where the full file passes 3/3? The PR-specific fix looks good from the targeted spec, but I would hesitate to mark this fully validated while the suite named in the PR body still fails locally.

**Comment by AmyShang-alt** [comment]
QA validation on head `578395750733863ea8896b19575bab7470c2d632`:

- PASS: new regression test `does not let a stalled attempt’s forced-shutdown timers kill the healthy retry`
- PASS: adjacent retry/lifecycle suites: `run-retry-policy`, `run-lifecycle-tracer`, `daemon-lifecycle` = 25/25
- PASS: `pnpm --filter @open-design/daemon run build`
- PASS: `pnpm --filter @open-design/daemon exec tsc -p tsconfig.json --noEmit`

Known non-blocking baseline issue: `tests/run-retry-runtime.test.ts -t "retries a transient first-token failure"` fails locally with `expected start events length 2, got 1`, but the same failure reproduces on current main baseline (`/Users/mac/open-design/main-5202-baseline`, commit `7b9864614`). Treating this as pre-existing/non-PR-blocking for #5202.

Conclusion: #5202 core watchdog cross-generation fix is validated. OK to remove `needs-validation`.

File: apps/daemon/tests/run-retry-runtime.test.ts (hunk 2 above)
Comment (AmyShang-alt) [reply apps/daemon/tests/run-retry-runtime.test.ts:202]: Follow-up: I reproduced the same `start events length 2, got 1` failure on current main baseline (`/Users/mac/open-design/main-5202-baseline`, commit `7b9864614`), so this is pre-existing and not blocking #5202. The PR-specific watchdog regression passes, adjacent retry/lifecycle suites plus daemon build/typecheck pass, and I have marked the PR validated.

File: apps/daemon/tests/run-retry-runtime.test.ts (hunk 1 above)
Comment (AmyShang-alt) [reply apps/daemon/tests/run-retry-runtime.test.ts:189]: Confirmed addressed by the follow-up commit: `OD_CHAT_RUN_INACTIVITY_KILL_GRACE_MS` is now included in the env snapshot/restore path, and adjacent retry/lifecycle suites passed during QA validation. Resolving.
